File: backend/internal/service/asset_changes.go
package service

import (
	"context"
	"encoding/json"
	"errors"
	"time"

	"photo-map-app/backend/internal/db/sqlc"
	"photo-map-app/backend/internal/model"
)

const (
	assetChangeUpsert  = "upsert"
	assetChangeTrash   = "trash"
	assetChangeRestore = "restore"
	assetChangeDelete  = "delete"

	defaultAssetChangesLimit = 500
	maxAssetChangesLimit     = 1000
)

type ListAssetChangesParams struct {
	UserID string
	Cursor int64
	Limit  int
}

func BuildAssetSnapshot(asset sqlc.Asset) model.AssetSnapshot {
	return model.AssetSnapshot{
		ID:                    asset.ID,
		MediaType:             asset.MediaType,
		MimeType:              asset.MimeType,
		OriginalFilename:      asset.OriginalFilename,
		FileSizeBytes:         asset.FileSizeBytes,
		ChecksumSha256:        asset.ChecksumSha256,
		ThumbnailKey:          asset.ThumbnailKey,
		PreviewKey:            asset.PreviewKey,
		PosterFrameKey:        asset.PosterFrameKey,
		TakenAt:               asset.TakenAt,
		TakenAtSource:         asset.TakenAtSource,
		TimezoneOffsetMinutes: asset.TimezoneOffsetMinutes,
		Width:                 asset.Width,
		Height:                asset.Height,
		DurationMs:            asset.DurationMs,
		Orientation:           asset.Orientation,
		Latitude:              asset.Latitude,
		Longitude:             asset.Longitude,
		Country:               asset.Country,
		Region:                asset.Region,
		City:                  asset.City,
		PlaceName:             asset.PlaceName,
		CameraMake:            asset.CameraMake,
		CameraModel:           asset.CameraModel,
		Software:              asset.Software,
		IsFavorite:            boolValue(asset.IsFavorite),
		IsArchived:            boolValue(asset.IsArchived),
		IsTrashed:             boolValue(asset.IsTrashed),
		UploadedAt:            asset.UploadedAt,
		UpdatedAt:             asset.UpdatedAt,
	}
}

func marshalAssetSnapshot(asset sqlc.Asset) ([]byte, error) {
	return json.Marshal(BuildAssetSnapshot(asset))
}

func insertAssetChange(
	ctx context.Context,
	queries *sqlc.Queries,
	asset sqlc.Asset,
	changeType string,
) (sqlc.AssetChange, error) {
	snapshot, err := marshalAssetSnapshot(asset)
	if err != nil {
		return sqlc.AssetChange{}, err
	}
	return queries.InsertAssetChange(ctx, sqlc.InsertAssetChangeParams{
		UserID:        asset.UserID,
		AssetID:       asset.ID,
		ChangeType:    changeType,
		AssetSnapshot: snapshot,
	})
}

func insertAssetDeleteChange(
	ctx context.Context,
	queries *sqlc.Queries,
	userID string,
	assetID string,
) (sqlc.AssetChange, error) {
	return queries.InsertAssetChange(ctx, sqlc.InsertAssetChangeParams{
		UserID: userID, ChangeType: assetChangeDelete, AssetID: assetID, AssetSnapshot: nil,
	})
}

func (service *AssetService) ListChanges(
	ctx context.Context,
	params ListAssetChangesParams,
) (model.AssetChangesResponse, error) {
	if params.Cursor < 0 {
		return model.AssetChangesResponse{}, ErrInvalidCursor
	}
	limit := normalizeAssetChangesLimit(params.Limit)
	changes, err := service.queries.ListAssetChangesAfterCursor(ctx, sqlc.ListAssetChangesAfterCursorParams{
		UserID: params.UserID,
		Cursor: params.Cursor,
		Limit:  int32(limit),
	})
	if err != nil {
		return model.AssetChangesResponse{}, err
	}

	items := make([]model.AssetChangeItem, 0, len(changes))
	nextCursor := params.Cursor
	for _, change := range changes {
		item, err := service.mapAssetChange(ctx, change)
		if err != nil {
			return model.AssetChangesResponse{}, err
		}
		items = append(items, item)
		nextCursor = change.ChangeID
	}

	serverCursor, err := service.queries.GetLatestAssetChangeIDForUser(ctx, params.UserID)
	if err != nil {
		return model.AssetChangesResponse{}, err
	}

	return model.AssetChangesResponse{
		Items:        items,
		NextCursor:   nextCursor,
		HasMore:      nextCursor < serverCursor,
		ServerCursor: serverCursor,
		ServerTime:   time.Now().UTC(),
	}, nil
}

func (service *AssetService) mapAssetChange(
	ctx context.Context,
	change sqlc.AssetChange,
) (model.AssetChangeItem, error) {
	item := model.AssetChangeItem{
		ChangeID: change.ChangeID, AssetID: change.AssetID,
		ChangeType: change.ChangeType, ChangedAt: change.ChangedAt,
	}
	if change.ChangeType == assetChangeDelete {
		return item, nil
	}
	if len(change.AssetSnapshot) == 0 {
		return model.AssetChangeItem{}, errors.New("asset change snapshot is missing")
	}

	var snapshot model.AssetSnapshot
	if err := json.Unmarshal(change.AssetSnapshot, &snapshot); err != nil {
		return model.AssetChangeItem{}, err
	}
	asset := model.AssetChangeAsset{AssetSnapshot: snapshot}
	var err error
	asset.ThumbnailURL, err = service.presignOptionalKey(ctx, snapshot.ThumbnailKey)
	if err != nil {
		return model.AssetChangeItem{}, err
	}
	asset.PreviewURL, err = service.presignOptionalKey(ctx, snapshot.PreviewKey)
	if err != nil {
		return model.AssetChangeItem{}, err
	}
	asset.PosterFrameURL, err = service.presignOptionalKey(ctx, snapshot.PosterFrameKey)
	if err != nil {
		return model.AssetChangeItem{}, err
	}
	item.Asset = &asset
	return item, nil
}

func (service *AssetService) presignOptionalKey(ctx context.Context, key *string) (*string, error) {
	if key == nil || *key == "" {
		return nil, nil
	}
	url, err := service.storageService.GeneratePresignedGetURL(ctx, *key, service.readExpires)
	if err != nil {
		return nil, err
	}
	return &url, nil
}

func normalizeAssetChangesLimit(limit int) int {
	if limit <= 0 {
		return defaultAssetChangesLimit
	}
	if limit > maxAssetChangesLimit {
		return maxAssetChangesLimit
	}
	return limit
}

File: web/src/features/assets/assetChanges.ts
import { appDb, type RemoteAssetRow } from "../../db/appDb";
import { apiRequest } from "../../lib/apiClient";
import {
  clearRemoteAssets,
  deleteRemoteAsset,
  putRemoteAsset,
  updateRemoteAsset,
} from "./assetReplica";

const ASSET_METADATA_STATE_KEY = "asset_metadata" as const;
const PAGE_LIMIT = 400;

export interface SyncAssetsChangesOptions {
  full?: boolean;
  onProgress?: (progress: AssetSyncProgress) => void;
}

export interface AssetSyncProgress {
  completedCount: number;
  remainingCount: number | null;
  totalCount: number | null;
  percent: number | null;
}

interface AssetSnapshot {
  id: string;
  mediaType: string;
  mimeType: string;
  originalFilename: string | null;
  fileSizeBytes: number;
  checksumSha256: string;
  thumbnailKey: string | null;
  previewKey: string | null;
  posterFrameKey: string | null;
  takenAt: string | null;
  takenAtSource: string | null;
  timezoneOffsetMinutes: number | null;
  width: number | null;
  height: number | null;
  durationMs: number | null;
  orientation: number | null;
  latitude: number | null;
  longitude: number | null;
  country: string | null;
  region: string | null;
  city: string | null;
  placeName: string | null;
  cameraMake: string | null;
  cameraModel: string | null;
  software: string | null;
  isFavorite: boolean;
  isArchived: boolean;
  isTrashed: boolean;
  uploadedAt: string;
  updatedAt: string;
}

interface AssetChangeAsset extends AssetSnapshot {
  // Signed URLs are temporary and may expire; keep them as transient cache fields.
  thumbnailUrl: string | null;
  previewUrl: string | null;
}

type AssetChangeType = "upsert" | "trash" | "restore" | "delete";

interface AssetChangeItem {
  changeId: number;
  assetId: string;
  changeType: AssetChangeType;
  changedAt: string;
  asset: AssetChangeAsset | null;
}

interface AssetChangesPage {
  items: AssetChangeItem[];
  nextCursor: number;
  hasMore: boolean;
  remainingCount?: number;
  serverCursor: number;
  serverTime: string;
}

export function calculateAssetSyncProgress(
  completedCount: number,
  remainingCount: number | undefined,
): AssetSyncProgress {
  if (remainingCount === undefined) {
    return {
      completedCount,
      remainingCount: null,
      totalCount: null,
      percent: null,
    };
  }

  const safeRemainingCount = Math.max(0, remainingCount);
  const totalCount = completedCount + safeRemainingCount;
  const percent =
    totalCount === 0 ? 100 : Math.floor((completedCount * 100) / totalCount);

  return {
    completedCount,
    remainingCount: safeRemainingCount,
    totalCount,
    percent,
  };
}

interface ReadUrlResponse {
  url: string;
}

export type AssetReadUrlVariant = "thumbnail" | "preview";

function toRemoteAssetRow(
  ownerUserId: string,
  asset: AssetChangeAsset,
  changeId: number,
): RemoteAssetRow {
  return {
    ownerUserId,
    id: asset.id,
    mediaType: asset.mediaType,
    mimeType: asset.mimeType,
    originalFilename: asset.originalFilename,
    fileSizeBytes: asset.fileSizeBytes,
    checksumSha256: asset.checksumSha256,
    thumbnailKey: asset.thumbnailKey,
    previewKey: asset.previewKey,
    posterFrameKey: asset.posterFrameKey,
    takenAt: asset.takenAt,
    takenAtSource: asset.takenAtSource,
    timezoneOffsetMinutes: asset.timezoneOffsetMinutes,
    width: asset.width,
    height: asset.height,
    durationMs: asset.durationMs,
    orientation: asset.orientation,
    latitude: asset.latitude,
    longitude: asset.longitude,
    country: asset.country,
    region: asset.region,
    city: asset.city,
    placeName: asset.placeName,
    cameraMake: asset.cameraMake,
    cameraModel: asset.cameraModel,
    software: asset.software,
    isFavorite: asset.isFavorite,
    isArchived: asset.isArchived,
    isTrashed: asset.isTrashed,
    uploadedAt: asset.uploadedAt,
    updatedAt: asset.updatedAt,
    thumbnailUrl: asset.thumbnailUrl,
    previewUrl: asset.previewUrl,
    lastChangeId: changeId,
  };
}

async function getLastCommittedCursor(ownerUserId: string): Promise<number> {
  const state = await appDb.remote_sync_state_by_user.get([
    ownerUserId,
    ASSET_METADATA_STATE_KEY,
  ]);
  return state?.value ?? 0;
}

async function fetchChangesPage(cursor: number): Promise<AssetChangesPage> {
  const query = new URLSearchParams({
    cursor: String(cursor),
    limit: String(PAGE_LIMIT),
  });

  return apiRequest<AssetChangesPage>(`/assets/changes?${query.toString()}`);
}

async function applyChangesPage(
  ownerUserId: string,
  page: AssetChangesPage,
): Promise<void> {
  await appDb.transaction(
    "rw",
    appDb.remote_assets_by_user,
    appDb.remote_sync_state_by_user,
    async () => {
      for (const item of page.items) {
        if (item.changeType === "delete" || item.asset === null) {
          await deleteRemoteAsset(ownerUserId, item.assetId);
          continue;
        }

        const row = toRemoteAssetRow(ownerUserId, item.asset, item.changeId);
        await putRemoteAsset(ownerUserId, row);
      }

      await appDb.remote_sync_state_by_user.put({
        ownerUserId,
        key: ASSET_METADATA_STATE_KEY,
        value: page.nextCursor,
        updatedAt: new Date().toISOString(),
      });
    },
  );
}

export async function syncAssetsChanges(
  ownerUserId: string,
  options: SyncAssetsChangesOptions = {},
): Promise<number> {
  if (options.full) {
    await appDb.transaction(
      "rw",
      appDb.remote_assets_by_user,
      appDb.remote_sync_state_by_user,
      async () => {
        await clearRemoteAssets(ownerUserId);
        await appDb.remote_sync_state_by_user.put({
          ownerUserId,
          key: ASSET_METADATA_STATE_KEY,
          value: 0,
          updatedAt: new Date().toISOString(),
        });
      },
    );
  }

  let cursor = await getLastCommittedCursor(ownerUserId);
  let completedCount = 0;

  while (true) {
    const page = await fetchChangesPage(cursor);

    await applyChangesPage(ownerUserId, page);
    cursor = page.nextCursor;
    completedCount += page.items.length;
    options.onProgress?.(
      calculateAssetSyncProgress(completedCount, page.remainingCount),
    );

    if (!page.hasMore) {
      break;
    }
  }

  return cursor;
}

export async function clearAssetsReplicationCache(ownerUserId: string): Promise<void> {
  await appDb.transaction(
    "rw",
    appDb.remote_assets_by_user,
    appDb.remote_sync_state_by_user,
    async () => {
      await clearRemoteAssets(ownerUserId);
      await appDb.remote_sync_state_by_user.delete([
        ownerUserId,
        ASSET_METADATA_STATE_KEY,
      ]);
    },
  );
}

export async function getAssetReplicaReadUrl(
  assetId: string,
  variant: AssetReadUrlVariant,
): Promise<string> {
  const query = new URLSearchParams({ variant });
  const response = await apiRequest<ReadUrlResponse>(
    `/assets/${assetId}/read-url?${query.toString()}`,
  );
  return response.url;
}

export async function refreshAssetReplicaUrl(
  ownerUserId: string,
  assetId: string,
  variant: AssetReadUrlVariant,
): Promise<void> {
  const nextUrl = await getAssetReplicaReadUrl(assetId, variant);

  if (variant === "thumbnail") {
    await updateRemoteAsset(ownerUserId, assetId, { thumbnailUrl: nextUrl });
    return;
  }

  await updateRemoteAsset(ownerUserId, assetId, { previewUrl: nextUrl });
}
